### PR TITLE
Allow `@JsonMerge` with Custom List

### PR DESCRIPTION
**Pull request by JooHyukKim**
To address #4783 

Hopefully it's okay to add to 2.18

### CODE REVIEW

**Comment by JooHyukKim** [comment]
Feel free to close! Wrote a PR in case we might want to add this to test cases,

**Comment by cowtowncoder** [comment]
As per my comments elsewhere, I don't think `mapper.readerForUpdating()` should be needed in this case. Unless existing merge tests has something that work around the issue, different from case here.

**Comment by JooHyukKim** [comment]
You are right. `@JsonMerge` with custom Collection type does not work.
Modified test accordingly. 

I will probably investigate how different between JDK/custom collection case differ, probably at below point to be specific.

image

**Comment by cowtowncoder** [comment]
Note `_mapAbstractCollectionType` simply maps things like `Collection`/`List` -> `ArrayList`, `Set` -> `LinkedHashSet` and so on, to have concrete type to deserialize.
Custom type wouldn't map. But we can't resolve it by mapping; rather would need to figure out how to postpone failure so that lack of deserializer would not trigger exception for this particular `@Merge` case. Not sure how, but that'd be the idea.

It might be tricky, come to think of it, as there's need for a deserializer to handle merging, I think. Construction of which should not fail unless... hmmh.

File: src/main/java/com/fasterxml/jackson/databind/deser/BasicDeserializerFactory.java [hunk 1]
```diff
@@ -827,9 +827,16 @@ public JsonDeserializer<?> createCollectionDeserializer(DeserializationContext c
                 if (implType == null) {
                     // [databind#292]: Actually, may be fine, but only if polymorphich deser enabled
                     if (type.getTypeHandler() == null) {
-                        throw new IllegalArgumentException("Cannot find a deserializer for non-concrete Collection type "+type);
+                        // [databind#4783] Since 2.19, Allow `@JsonMerge` with Custom Collection extension.
+                        // Let's not throw exceptino so instantiation can happen downstream
+                        if (_canMapSuperInterfaceAsAbstractCollectionType(type, config)) {
```
Comment (JooHyukKim) [inline src/main/java/com/fasterxml/jackson/databind/deser/BasicDeserializerFactory.java:832]: .

File: src/test/java/com/fasterxml/jackson/databind/tofix/JsonMergeWithCustomCollection4783Test.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+import com.fasterxml.jackson.databind.json.JsonMapper;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+// [databind#4783] Test to verify that JsonMerge also works for custom list
+public class JsonMergeWithCustomCollection4783Test
+{
+
+    private static class MyArrayListJDK<T> extends ArrayList<T> implements List<T> { }
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/tofix/JsonMergeWithCustomCollection4783Test.java:20]: No need for `implements List<T>`, already does via `ArrayList<T>`.

File: src/test/java/com/fasterxml/jackson/databind/tofix/JsonMergeWithCustomCollection4783Test.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    interface MyListCustom<T> extends List<T> { }
+
+    private static class MyArrayListCustom<T> extends ArrayList<T> implements MyListCustom<T> { }
+
+    static class MergeCustomStringList {
+        @JsonMerge
+        @JsonProperty
+        public MyListCustom<String> values = createCustomStringList();
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/tofix/JsonMergeWithCustomCollection4783Test.java:40]: I don't think helper method needed? Just create directly with `new MyArrayListCustom<>()`

File: src/test/java/com/fasterxml/jackson/databind/tofix/JsonMergeWithCustomCollection4783Test.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        { values.add("a"); }
+    }
+
+    static <T> MyListCustom<T> createCustomStringList() {
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/tofix/JsonMergeWithCustomCollection4783Test.java:44]: Not really needed as per above (but if it was, should parameterize with `String` to align with name.

File: src/test/java/com/fasterxml/jackson/databind/tofix/JsonMergeWithCustomCollection4783Test.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    interface MyCustomLongList<T> extends List<T> { }
+
+    private static class MyArrayCustomLongList<T> extends ArrayList<T> implements MyCustomLongList<T> { }
+
+    static class MergeMyCustomLongList {
+        @JsonMerge
+        @JsonProperty
+        public MyCustomLongList<String> values = createCustomLongList();
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/tofix/JsonMergeWithCustomCollection4783Test.java:55]: Hmmmh. If its "Long" list, shouldn't it contain `Long` values?

File: src/test/java/com/fasterxml/jackson/databind/tofix/JsonMergeWithCustomCollection4783Test.java (hunk 3 above)
Comment (cowtowncoder) [reply src/test/java/com/fasterxml/jackson/databind/tofix/JsonMergeWithCustomCollection4783Test.java:40]: Unless you want to add values in this method?

**Review by cowtowncoder — COMMENTED** [review]
Ok, so, I think we can simplify the fix by using `AbstractDeserializer` for all cases, and see how that works.

**Comment by cowtowncoder** [comment]
> Took a shot at solving this issue. Added [comment🔗](https://github.com/FasterXML/jackson-databind/pull/4784/files#r1834435293) at the implementation. PTAL when you have time? @cowtowncoder
> 
> PS: Looking at current implementation I am not sure if we have this merged to 2.18 (seems a bit risky?)

I agree, bit risky for 2.18, based on experience. So let's target 2.19

File: src/main/java/com/fasterxml/jackson/databind/deser/BasicDeserializerFactory.java (hunk 1 above)
Comment (JooHyukKim) [reply src/main/java/com/fasterxml/jackson/databind/deser/BasicDeserializerFactory.java:832]: I tried actually, it fails with below exception.
I checked what gets created inside `AbstractDeserializer` and almost nothing to actually deserialize anything.
The intent behind the `// do nothing` current approach was to create  `StringCollectionDeserializer` / `CollectionDeserializer` below starting at line 862.

```text
com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `com.fasterxml.jackson.databind.tofix.JsonMergeWithCustomCollection4783Test$MyListCustom` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 11] (through reference chain: com.fasterxml.jackson.databind.tofix.JsonMergeWithCustomCollection4783Test$MergeCustomStrin
	```

Comment (JooHyukKim) [reply src/main/java/com/fasterxml/jackson/databind/deser/BasicDeserializerFactory.java:832]: But wouldn't hurt showcasing here! Lemme just make the suggested change so we both can see

Comment (JooHyukKim) [reply src/main/java/com/fasterxml/jackson/databind/deser/BasicDeserializerFactory.java]: Yup, you can see checks are failing with the same exceptions.

**Comment by JooHyukKim** [comment]
> EDIT: I notice pr is against 2.18. We can consider that if necessary.

I didn't think this was an urgent issue actually.
I was even thinking to close(to postpone) this one since you got so much going on now @cowtowncoder

**Comment by cowtowncoder** [comment]
@JooHyukKim Ok thanks. I think we are good -- let's try getting this done, I think it is a valid problem to solve. I'll see why the test failures occur...

**Comment by cowtowncoder** [comment]
Ahhh. Ok, so that'd be too easy -- things won't work because `AbstractDeserializer` does not override 

    public T deserialize(JsonParser p, DeserializationContext ctxt, T intoValue)

and it can't really, since we need method like one from `CollectionDeserializer`

    @Override
    public Collection<Object> deserialize(JsonParser p, DeserializationContext ctxt,
            Collection<Object> result)

for merging to actually work. So basically we need a proper `CollectionDeserializer` that... hmmh. 

Let me try a simple change to see what happens if we simply remove the check for abstract altogether.

**Comment by cowtowncoder** [comment]
Well look at that -- tests actually pass. But probably need a test for non-merging case, to see how error handling works (when trying to deserialize abstract collection without merging)

**Comment by cowtowncoder** [comment]
@JooHyukKim Ok so it appears fix might be surprisingly simple: just basically remove the "is it abstract -> fail" check. Love when this happens. :)

**Comment by cowtowncoder** [comment]
Due to simplicity of the fix, I think 2.18 might be fine after all.

**Comment by JooHyukKim** [comment]
> @JooHyukKim Ok so it appears fix might be surprisingly simple: just basically remove the "is it abstract -> fail" check. Love when this happens. :)

Interestinnggg :)))
Thank you for the improvement with additional test and all 👌🏼👌🏼
Fixed test class name typo, added release notes and such.

**Comment by cowtowncoder** [comment]
Minor tweak: remove change wrt Map defaulting: will do that for 2.19, to reduce risk for 2.18.2.

**Pull request merged**